File: zig-impl/src/days/day01.zig
const std = @import("std");
const utils = @import("./../utils/utils.zig");

pub fn part1(allocator: std.mem.Allocator, lines: [][]const u8) !i64 {
    _ = allocator;
    var tracker: i32 = 50;
    var password: i32 = 0;
    for (lines) |line| {
        if (line.len == 0) continue;
        const direction = line[0];
        const rotations = try std.fmt.parseInt(i32, line[1..], 10);
        switch (direction) {
            'R' => tracker = @mod(tracker + rotations, 100),
            'L' => tracker = @mod(tracker - rotations, 100),
            else => {},
        }
        if (tracker == 0) {
            password += 1;
        }
    }
    return password;
}

pub fn part2(allocator: std.mem.Allocator, lines: [][]const u8) !i64 {
    _ = allocator;
    var dial: i32 = 50;
    var password: i32 = 0;
    for (lines) |line| {
        // std.debug.print("\n Instruction:{s} Dial:{} Password:{}", .{ line, dial, password });
        if (line.len == 0) continue;
        const direction = line[0];
        const rotations = try std.fmt.parseInt(i32, line[1..], 10);
        switch (direction) {
            'R' => {
                password += @divTrunc(dial + rotations, 100);
                dial = @mod(dial + rotations, 100);
            },
            'L' => {
                const new_dial = @mod(dial - rotations, 100);
                if (dial == 0) {
                    password += @divFloor(rotations, 100);
                } else if (rotations > dial) {
                    password += @divFloor((rotations - dial - 1), 100) + 1;
                    if (new_dial == 0) {
                        password += 1;
                    }
                } else if (dial - rotations == 0) {
                    password += 1;
                }
                dial = new_dial;
            },
            else => {},
        }
    }
    return password;
}

test "part1 output test" {
    var gpa = std.heap.GeneralPurposeAllocator(.{ .thread_safe = true }){};
    const allocator = gpa.allocator();
    const input =
        \\L68
        \\L30
        \\R48
        \\L5
        \\R60
        \\L55
        \\L1
        \\L99
        \\R14
        \\L82
    ;
    var lines = try utils.parseLines(allocator, input);
    defer lines.deinit(allocator);

    const result = try part1(allocator, lines.items);
    try std.testing.expectEqual(3, result);
}

test "part2 output test" {
    var gpa = std.heap.GeneralPurposeAllocator(.{ .thread_safe = true }){};
    const allocator = gpa.allocator();
    const input =
        \\L68
        \\L30
        \\R48
        \\L5
        \\R60
        \\L55
        \\L1
        \\L99
        \\R14
        \\L82
    ;
    var lines = try utils.parseLines(allocator, input);
    defer lines.deinit(allocator);

    const result = try part2(allocator, lines.items);
    try std.testing.expectEqual(6, result);
}

File: python-impl/days/day01.py
def part1(lines: list[str]) -> int:
    tracker = 50
    password = 0

    for line in lines:
        if len(line) == 0:
            continue

        direction = line[0]
        rotations = int(line[1:])

        if direction == "R":
            tracker = (tracker + rotations) % 100
        elif direction == "L":
            tracker = (tracker - rotations) % 100

        if tracker == 0:
            password += 1

    return password


def part2(lines: list[str]) -> int:
    dial = 50
    password = 0

    for line in lines:
        if len(line) == 0:
            continue

        direction = line[0]
        rotations = int(line[1:])

        if direction == "R":
            password += (dial + rotations) // 100
            dial = (dial + rotations) % 100
        elif direction == "L":
            new_dial = (dial - rotations) % 100

            if dial == 0:
                password += rotations // 100
            elif rotations > dial:
                password += ((rotations - dial - 1) // 100) + 1
                if new_dial == 0:
                    password += 1
            elif dial - rotations == 0:
                password += 1

    return password


def test_part1():
    input_data = """L68
L30
R48
L5
R60
L55
L1
L99
R14
L82"""
    lines = input_data.strip().split("\n")
    result = part1(lines)
    assert result == 3, f"Expected 3, got {result}"
    print(f"Part 1 test passed: {result}")


def test_part2():
    input_data = """L68
L30
R48
L5
R60
L55
L1
L99
R14
L82"""
    lines = input_data.strip().split("\n")
    result = part2(lines)
    assert result == 6, f"Expected 6, got {result}"
    print(f"Part 2 test passed: {result}")


if __name__ == "__main__":
    test_part1()
    test_part2()
    print("All tests passed!")
